### PR TITLE
TRITON-2228: Linux CN minimum viable product

### PR DESCRIPTION
**Pull request by bahamat**

### CODE REVIEW

File: lib/algorithms/hard-filter-os.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	assert.arrayOfObject(servers, 'servers');
+	assert.object(opts, 'opts');
+	assert.object(opts.vm, 'opts.vm');
+	assert.string(opts.vm.brand, 'opts.vm.brand');
+	assert.optionalNumber(opts.vm.vcpus, 'opts.vm.vcpus');
+	assert.func(cb, 'cb');
+
+	var reasons = {};
+	var imgType = opts.img && opts.img.type;
+	var requireLinux = opts.vm.brand === 'lx' && imgType === 'lxd';
```
Comment (danmcd) [inline lib/algorithms/hard-filter-os.js:32]: Readability nit: parenthesize?
    var requireLinux = (opts.vm.brand === 'lx' && imgType === 'lxd');

**Review by danmcd — COMMENTED** [review]
One nit so far...

**Comment by danmcd** [comment]
Also, it appears all of your test cases are failure ones.  Can/should you have any success ones as well (i.e. some with both a Linux and an illumos server?)

File: lib/algorithms/hard-filter-os.js (hunk 1 above)
Comment (bahamat) [reply lib/algorithms/hard-filter-os.js]: Fixed.

File: test/algorithms/hard-filter-os.test.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+
+test('filterOS() one linux server', function (t) {
+	var servers = [
+		{
+			sysinfo: {},
+			uuid: 'f667e0fa-33db-48da-a5d0-9fe837ce93fc'
+		},
+		{
+			sysinfo: {'System Type': 'linux'},
```
Comment (danmcd) [inline test/algorithms/hard-filter-os.test.js:83]: Does this now need to be capitalized as well?

File: test/algorithms/hard-filter-os.test.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+
+test('filterOS() one linux server (lxd image)', function (t) {
+	var servers = [
+		{
+			sysinfo: {},
+			uuid: 'f667e0fa-33db-48da-a5d0-9fe837ce93fc'
+		},
+		{
+			sysinfo: {'System Type': 'linux'},
```
Comment (danmcd) [inline test/algorithms/hard-filter-os.test.js:112]: Same here?

**Review by danmcd — COMMENTED** [review]
I think you need to update your tests to use 'Linux' as well.  Also, you should put 1-2 success tests in there (with an illumos CN and a Linux CN).

File: test/algorithms/hard-filter-os.test.js (hunk 2 above)
Comment (bahamat) [reply test/algorithms/hard-filter-os.test.js:83]: Ah, yes. Thanks.

**Comment by twhiteman** [comment]
> Also, it appears all of your test cases are failure ones. Can/should you have any success ones as well (i.e. some with both a Linux and an illumos server?)

Actually - 3 of the 4 tests do include successful server match cases (you can see via the `expectServers = ...`) - which also cover both Linux and SmartOS servers (SmartOS is the default when the server does not provide a 'System Type' via sysinfo).

**Comment by danmcd** [comment]
> (SmartOS is the default when the server does not provide a 'System Type' via sysinfo).

I should've paid more attention to the code, then.  That's on me.

**Review by danmcd — DISMISSED** [review]
Ship it.

**Comment by bahamat** [comment]
All tests now pass

```
1..1567
# tests 1567
# pass  1567

# ok
```

**Review by danmcd — APPROVED** [review]

**Pull request merged**